core/Symbols.cc

```diff
@@ -62,18 +62,39 @@ TypePtr Symbol::externalType(const GlobalState &gs) const {
             newResultType = make_type<ClassType>(ref);
         } else {
             vector<TypePtr> targs;
-            for (auto tm : typeMembers()) {
+            targs.reserve(typeMembers().size());
+
+            // Special-case covariant stdlib generics to have their types
+            // defaulted to `T.untyped`. This set *should not* grow over time.
+            bool isStdlibGeneric = ref == core::Symbols::Hash() || ref == core::Symbols::Array() ||
```

Comment:
What's your current feel about this. Do we want to reduce this list?
If we don't, I'm concerned that the alternative code path won't be tested enough as majority of commonly used generics wont take it.

Agreed on separating that piece of work

Reply:
I think it would be great to reduce the list, but there are some strange interactions with our existing covariant datatypes. For example, `ARGV` has the type `T::Array[T.untyped]`, and when you ask if it `is_a? Array`, the resulting type you get is `T::Array[<any>]`. It makes sense why this happens (`T.untyped` is an identity element for `glb`), but I'm not sure it's desirable.

A simple fix here would be to update the stdlib rbis, but given that we'll generate rbis for other libraries that have types like `T::Array[T.untyped]` in them, I'm not sure if removing the special case is the best path forward.

Reply:
isn't `<any>` also an identity?
Is this a question of reordering a a few`if`s?

Reply:
We would need to make the identity cases a little more complicated, but we could get them to prefer `T.untyped` when the two types passed in are `T.untyped` and `<any>`.

Reply:
I think I'm going to leave the special case in for now, and make an issue about revisiting it in the future.